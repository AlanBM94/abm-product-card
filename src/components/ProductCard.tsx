import React, { createContext } from 'react';
import styles from './../styles/styles.module.css';
import useProductCard from '../hooks/useProductCard';
import {
  ProductCardProps,
  ProductContextProps,
} from '../interfaces/interfaces';

export const ProductContext = createContext({} as ProductContextProps);

const { Provider } = ProductContext;

export const ProductCard = ({
  children,
  product,
  className,
  style,
  onChange,
  value,
  initialValues,
}: ProductCardProps) => {
  const {
    counter,
    increaseBy,
    maxCount,
    isMaxCountReached,
    reset,
  } = useProductCard({
    onChange,
    product,
    value,
    initialValues,
  });

  return (
    <Provider
      value={{
        product,
        counter,
        increaseBy,
        maxCount,
      }}
    >
      <div style={style} className={`${styles.productCard} ${className}`}>
        {children({
          count: counter,
          isMaxCountReached,
          maxCount: initialValues?.maxCount,
          product,
          increaseBy,
          reset,
        })}
      </div>
    </Provider>
  );
};
