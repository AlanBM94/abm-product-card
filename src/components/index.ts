import { ProductCardProps } from "../interfaces/interfaces";
export { ProductButtons } from "./ProductButtons";
export { ProductImage } from "./ProductImage";
export { ProductTitle } from "./ProductTitle";
export { ProductCard } from "./ProductCard";

export interface ProductCardHOCProps {
  ({ children, product }: ProductCardProps): JSX.Element;
  Title: ({ title }: { title?: string }) => JSX.Element;
  Image: ({ img }: { img?: string }) => JSX.Element;
  Buttons: () => JSX.Element;
}
