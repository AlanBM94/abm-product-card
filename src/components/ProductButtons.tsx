import React, { useCallback, useContext } from 'react';
import { ProductContext } from './ProductCard';

import styles from './../styles/styles.module.css';

export const ProductButtons = () => {
  const { increaseBy, counter, maxCount } = useContext(ProductContext);

  const isMaxReached = useCallback(() => {
    return !!maxCount && counter === maxCount;
  }, [counter, maxCount]);

  return (
    <div className={styles.buttonsContainer}>
      <button onClick={() => increaseBy(-1)} className={styles.buttonMinus}>
        -
      </button>
      <div className={styles.countLabel}>{counter}</div>
      <button
        onClick={() => increaseBy(+1)}
        className={`${styles.buttonAdd} ${isMaxReached() && styles.disabled}`}
      >
        +
      </button>
    </div>
  );
};
