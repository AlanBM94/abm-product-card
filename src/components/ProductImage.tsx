import React, { useContext } from 'react';
import { ProductContext } from './ProductCard';

import styles from './../styles/styles.module.css';
import noImage from './../assets/no-image.jpg';

export const ProductImage = ({ img = '' }: { img?: string }) => {
  const { product } = useContext(ProductContext);

  let imgToShow: string;

  if (img) {
    imgToShow = img;
  } else if (product.img) {
    imgToShow = noImage;
  } else {
    imgToShow = noImage;
  }

  return (
    <img src={imgToShow} alt="Product image" className={styles.productImg} />
  );
};
