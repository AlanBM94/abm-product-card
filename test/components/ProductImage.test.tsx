import React from 'react';
import renderer from 'react-test-renderer';
import { ProductImage, ProductCard } from '../../src/components/';
import { product } from '../data/products';
import 'jest-styled-components';

describe('Product image', () => {
  test('must show the component with a custom image', () => {
    const wrapper = renderer.create(<ProductImage img="https://hola.jpg" />);
    expect(wrapper.toJSON()).toMatchSnapshot();
  });

  test('must show the component with product name', () => {
    const wrapper = renderer.create(
      <ProductCard product={product}>{() => <ProductImage />}</ProductCard>
    );
    expect(wrapper.toJSON()).toMatchSnapshot();
  });
});
