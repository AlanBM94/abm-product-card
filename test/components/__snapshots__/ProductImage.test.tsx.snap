// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Product image must show the component with a custom image 1`] = `
<img
  alt="Product image"
  className="productImg"
  src="https://hola.jpg"
/>
`;
